packages/preferences/src/browser/folders-preferences-provider.ts

```diff
@@ -21,87 +21,113 @@ import { FolderPreferenceProvider, FolderPreferenceProviderFactory } from './fol
 import { FileSystem, FileStat } from '@theia/filesystem/lib/common';
 import URI from '@theia/core/lib/common/uri';
 
+export const LAUNCH_PROPERTY_NAME = 'launch';
+export type ResourceKind = 'settings' | 'launch';
+
 @injectable()
 export class FoldersPreferencesProvider extends PreferenceProvider {
 
     @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
     @inject(FileSystem) protected readonly fileSystem: FileSystem;
     @inject(FolderPreferenceProviderFactory) protected readonly folderPreferenceProviderFactory: FolderPreferenceProviderFactory;
 
-    private providers: FolderPreferenceProvider[] = [];
+    private providersByKind: Map<ResourceKind, FolderPreferenceProvider[]> = new Map();
+    private resourceKinds: ResourceKind[] = ['launch', 'settings'];
 
     @postConstruct()
     protected async init(): Promise<void> {
         await this.workspaceService.roots;
```

Comment:
I've started rewriting this code. It's unnecessary complicated.

Reply:
I've started reworking code as we decided earlier, so it will be much simpler.